### PR TITLE
Layers visibility

### LINKED ISSUE #11646: Toggle the visibility of layers
## Feature Description

Users have indicated that they are looking forward to the ability to enable/disable layer visibility on demand in the editor, similar to how Figma & Photoshop are able to do this.

### PR DESCRIPTION
**Pull request by merapi**
## Summary

image

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

If you hide the layer, it should remain hidden everywhere (page, extra pages on the side, thumbnails, preview/output etc.)

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11646

### CODE REVIEW

File: packages/story-editor/src/utils/storyPageToNode.js [hunk 1]
```diff
@@ -92,13 +92,15 @@ const PageWithDependencies = forwardRef(function PageWithDependencies(
...
-                    <DisplayElement
-                      key={element.id}
-                      previewMode
-                      element={element}
-                    />
-                  ))}
+                  {page.elements
+                    .filter((element) => !element.isHidden)
+                    .map((element) => (
+                      <DisplayElement
```
Comment (miina) [inline packages/story-editor/src/utils/storyPageToNode.js]: Maybe it would be simpler to just add this to the `DisplayElement` directly so anywhere its used it'll be hidden?

File: packages/story-editor/src/components/panels/layer/groupLayer.js [hunk 2]
```diff
@@ -111,8 +113,10 @@ function GroupLayer({ groupId }) {
       }
       layerName={name}
       handleClick={handleClick}
+      hasLayerHiddenIcon={isHidden}
       isSelected={isSelected}
       LayerIcon={GroupLayerIcon}
+      LayerVisibilityIcon={LayerVisibilityIcon}
```
Comment (miina) [inline packages/story-editor/src/components/panels/layer/groupLayer.js:125]: Q: Is there a benefit to passing this forward as a prop instead of using `hasLayerHiddenIcon` in `Layer` component to get the `LayerVisibilityIcon`?

File: packages/story-editor/src/app/rightClickMenu/items/layerHide.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+function LayerHide() {
+  const updateSelectedElements = useStory(
+    (ctx) => ctx.actions.updateSelectedElements
+  );
+  const toggleLayerHide = useCallback(
+    () =>
+      updateSelectedElements({
+        properties: (oldElement) => ({
+          ...oldElement,
+          isHidden: !oldElement.isHidden,
```
Comment (miina) [inline packages/story-editor/src/app/rightClickMenu/items/layerHide.js:38]: Should we deselect the element when hiding?

**Review by miina — COMMENTED** [review]
Looks good!

I can see that it's possible to edit an element when the layer is hidden (e.g. change the position) but not possible to move via keyboard / mouse. Is that intentional?

One more note: when a group is hidden but an element inside it not, the element is still visible. Should the whole group get hidden?

**Comment by merapi** [comment]
> I can see that it's possible to edit an element when the layer is hidden (e.g. change the position) but not possible to move via keyboard / mouse. Is that intentional?

You can move the hidden element via keyboard, but because the canvas focus is buggy(?) it's not easy to get there (select the element in the layers panel, click floating menu, press ESC, now you can move). In fact, you cannot move any (normal) element that way, from the layers panel.

> One more note: when a group is hidden but an element inside it not, the element is still visible. Should the whole group get hidden?

If you hide the group but then show one of the elements it will be visible and we don't have any mixed state for the whole group (no third icon), so the group visibility icon acts as a toggle for the elements inside it (set them once to hidden or visible).

File: packages/story-editor/src/app/rightClickMenu/items/layerHide.js (hunk 3 above)
Comment (merapi) [reply packages/story-editor/src/app/rightClickMenu/items/layerHide.js:38]: I have no strong feeling about it, it makes sense because what more do you want to do with it (tho you can still move it via keyboard or edit via Style tab), but you also lose the track of your position in the layers panel, so when I just want to hide the layer temporarily and show it after I get the clear view of the canvas - it will not be helpful for me.

File: packages/story-editor/src/components/panels/layer/groupLayer.js (hunk 2 above)
Comment (merapi) [reply packages/story-editor/src/components/panels/layer/groupLayer.js:125]: I think no, but I did the same thing as we did for locking with hope that it makes sense 😁

File: packages/story-editor/src/utils/storyPageToNode.js (hunk 1 above)
Comment (merapi) [reply packages/story-editor/src/utils/storyPageToNode.js]: 👍🏻 , changed.

File: packages/story-editor/src/components/canvas/displayElement.js [hunk 4]
```diff
@@ -112,6 +112,10 @@ function DisplayElement({
   isAnimatable = false,
   siblingCount = 0,
 }) {
+  if (element.isHidden) {
+    return null;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/displayElement.js]: This has to be moved to after the hooks - or the component should be refactored into two, to short-circuit in case of hidden element.

File: packages/story-editor/src/components/canvas/frameElement.js [hunk 5]
```diff
@@ -153,11 +153,11 @@ function FrameElement({ id }) {
         isActive,
       };
     });
-  const { type, flip, isLocked } = element;
+  const { type, flip, isLocked, isHidden } = element;
 
   // Unlocked elements are always clickable,
   // locked elements are only clickable if selected
-  const isClickable = !isLocked || isSelected;
+  const isClickable = (!isLocked || isSelected) && !isHidden;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/frameElement.js]: Maybe add `// hidden elements are never clickable` to keep the comment updated.

File: packages/story-editor/src/components/canvas/selectionCanvas.js [hunk 6]
```diff
@@ -182,7 +182,10 @@ function SelectionCanvas({ children }) {
       const lassoP = new SAT.Box(new SAT.Vector(lx, ly), lw, lh).toPolygon();
       const newSelectedElementIds = currentPageElements
         // Skip background and locked elements
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/selectionCanvas.js]: Also update here

File: packages/story-editor/src/components/canvas/singleSelectionMoveable/index.js [hunk 7]
```diff
@@ -62,8 +62,9 @@ const SingleSelectionMoveable = forwardRef(function SingleSelectionMoveable(
     scrollTop: state.scrollTop,
   }));
 
-  const { isLocked } = selectedElement;
-  const actionsEnabled = !selectedElement.isBackground && !isLocked;
+  const { isLocked, isHidden } = selectedElement;
+  const actionsEnabled =
+    !selectedElement.isBackground && !isLocked && !isHidden;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/singleSelectionMoveable/index.js]: Why is `isBackground` not destructured? Looks weird.

File: packages/story-editor/src/components/panels/layer/groupLayerActions.js [hunk 8]
```diff
@@ -51,11 +51,25 @@ function GroupLayerActions({ groupId }) {
...
+
+  const handleHideGroup = useCallback(() => {
+    updateGroupById({
+      groupId,
+      properties: { isHidden: !isHidden },
+    });
+    updateElementsById({
+      elementIds: groupLayers.map((layer) => layer.id),
+      properties: { isHidden: !isHidden },
+    });
```
Comment (barklund) [inline packages/story-editor/src/components/panels/layer/groupLayerActions.js]: Hmm, do we really want to both keep hidden on the layer and the group level? If I hide all elements in a group one-by-one, shouldn't the group then appear hidden? And if I hide a group, unhide each layer in it, then the group appears hidden, but none of the layers are. If we only have hidden on the layer level, we can deduce it on the group level (though it requires some lookups).
Screen Shot 2022-11-28 at 10 58 03

File: packages/story-editor/src/components/panels/layer/layerComponents.js [hunk 9]
```diff
@@ -204,6 +217,9 @@ export const IconWrapper = styled.div`
     color: ${({ theme }) => theme.colors.fg.secondary};
   }
 `;
+export const HiddenIconWrapper = styled(IconWrapper)`
+  right: 78px;
+`;
```
Comment (barklund) [inline packages/story-editor/src/components/panels/layer/layerComponents.js]: I'm not sure I like the location of this one. Looks weird with a long layer name, that the layer name reappears on the other side of the icon. 

Screen Shot 2022-11-28 at 11 01 43

Maybe we should move all the sticky-able icons to the far right and if either sticky icon is shown, cut the text off at the same location.

This is what Figma does:

Screen Shot 2022-11-28 at 11 05 28

Also on Figma, the text is cut off by making the text field shorter (preserving ellipsis) rather than overlay (which we do).  We do have a slightly longer actions overlay though, so not sure if it's feasible. Also, would be great if we could reuse the actions component for the sticky components, now that we have two of them. It's a bit much duplication and comes with alignment issues.

**Review by barklund — APPROVED** [review]
Forgot to press submit! 🤦

File: packages/story-editor/src/components/canvas/displayElement.js (hunk 4 above)
Comment (merapi) [reply packages/story-editor/src/components/canvas/displayElement.js]: I was thinking about _Rules of Hooks_ when I added that, but it doesn't throw any warning - is this a special case when you can do the early return?

File: packages/story-editor/src/components/panels/layer/groupLayerActions.js (hunk 8 above)
Comment (merapi) [reply packages/story-editor/src/components/panels/layer/groupLayerActions.js]: Since we don't have this third state icon (like in Figma) we can deduce it on the group level as you propose.

File: packages/story-editor/src/components/canvas/displayElement.js (hunk 4 above)
Comment (barklund) [reply packages/story-editor/src/components/canvas/displayElement.js]: Nope, you can't do it here either. I don't know why you didn't see a warning, is clearly reported in CI, no?

File: packages/story-editor/src/components/panels/layer/groupLayer.js (hunk 2 above)
Comment (miina) [reply packages/story-editor/src/components/panels/layer/groupLayer.js:125]: Move both to reduce prop drilling without a reason? 😄

File: packages/story-editor/src/components/canvas/displayElement.js (hunk 4 above)
Comment (miina) [reply packages/story-editor/src/components/canvas/displayElement.js]: Yep, it's reported in CI.

**Review by miina — APPROVED** [review]
LGTM (pending the early `return null` issue)

File: packages/story-editor/src/components/canvas/displayElement.js (hunk 4 above)
Comment (merapi) [reply packages/story-editor/src/components/canvas/displayElement.js]: Good to hear that everything is as it should be, I'm blind 🤣. Fixed.

File: packages/story-editor/src/components/panels/layer/groupLayer.js (hunk 2 above)
Comment (merapi) [reply packages/story-editor/src/components/panels/layer/groupLayer.js:125]: Ok, so there is a reason to pass the Icon as a prop: the Layer component is used for both layer and group, for layer it's simple but for a group there is a condition, so we cannot hardcode the icon in the Layer component. Also, with the new layer visibility icon for the group (deduced from layers within the group) it makes sense to do it that way.

**Comment by swissspidy** [comment]
FYI there's a minor lint issue

File: packages/story-editor/src/components/panels/layer/layerComponents.js (hunk 9 above)
Comment (merapi) [reply packages/story-editor/src/components/panels/layer/layerComponents.js]: New version:
image

**Comment by merapi** [comment]
> Also, is there a follow-up ticket for adding tests?

I'm adding them here.

File: packages/story-editor/src/components/panels/layer/groupLayerActions.js (hunk 8 above)
Comment (merapi) [reply packages/story-editor/src/components/panels/layer/groupLayerActions.js]: Done.

File: packages/story-editor/src/components/panels/layer/layerComponents.js (hunk 9 above)
Comment (barklund) [reply packages/story-editor/src/components/panels/layer/layerComponents.js]: Looks much better!

**Comment by kkalarickal** [comment]
@merapi - I tested the layer hiding functionality and found one case where it is not working: 

1. add two products to a page and put them into a group
2. both products (i.e. hide the entire group) -> notice how the preview/published story does not show the products
3. now unhide one product but keep the other one hidden
Issue: Now the preview/published story still shows BOTH products
Expected: the product that was hidden in the layers config, should NOT show up in the preview or published story

Screenshot - Yellow Sofa is shows, but Vanilla Candle is hidden



Screenshot - both products are seen in the front end

**Comment by kkalarickal** [comment]
One more side effect of the grouping/layer hiding is that

1. add an image to the page and set it as background - note that it is not possible to hide this image
2. detach the image from background and include it in a group with some other elements
3. now hide the image within the group
4. right click and set the hidden image as background

Issue: It is possible to apply a hidden image as the background while it remains within the group

Side Effects:
- the preview of the story shows some weird color background instead of 
- while it is hidden, no other image can be applied as background

Screenshots - background image within a group
 

Screenshot - its possible to add a hidden image as background

**Comment by kkalarickal** [comment]
@merapi - apart from this, I hit a crash when trying to open/close the group within the layers config. Unfortunately, it was a one off thing and I could not reproduce it consistently. I will try again and see if I can get it to happen.

I also tested Firefox, Safari, Opera, Edge and confirmed that the layer hiding/unhiding works as expected.

Cc: @swissspidy

**Comment by merapi** [comment]
> @merapi - I tested the layer hiding functionality and found one case where it is not working:

Good find, indeed the product was not hidden in the Shop now/featured list. Fixed.

**Comment by merapi** [comment]
> add an image to the page and set it as background - note that it is not possible to hide this image

Until this point, it was as designed, but on the BB we decided it would also be nice to hide the background.
Working on it!

**Comment by merapi** [comment]
> Until this point, it was as designed, but on the BB we decided it would also be nice to hide the background.
Working on it!

I made some additional fixes from BB and merged the latest main, the only remaining thing is
`Question: should one be able to hide the background element? - need confirmation from Andrew` - there are two parts to this:
1) hide/show the background in the layers panel - pretty straightforward
2) output - how it should behave if there is a hidden background image/video/gif with or without background color because right now if you set the background video and hide it, it will not output a white background (like you see in the editor) it also doesn't account for the background color

Maybe it's a good idea to confirm this feature and output behavior with UX.

**Comment by kkalarickal** [comment]
> Good find, indeed the product was not hidden in the Shop now/featured list. Fixed.

Confirmed that the Product elements that are added to a page can be hidden individually or as a group and they don't show up in the preview, or after publishing.

> It is possible to hide an image asset and then apply it as background. But there is a weird color to the background when this happens

This is still an issue. Setting a hidden image asset as background leaves the preview or the published story with a background color that is not expected. Note, in the editor itself the background is white

Editor


Preview


Published story


Cc: @swissspidy @bmattb

**Comment by kkalarickal** [comment]
@swissspidy - the issue that remains is a pretty corner case. I feel we should open a sub-task and push this PR through since it implements the overall hiding of layers as expected.

**Comment by swissspidy** [comment]
IDK how you were able to set a hidden layer as the background element in the first place, but I created a ticket now to capture this, since it requires UX feedback: #12778
